test/extensions/filters/network/zookeeper_proxy/zookeeper_filter_test.cc

```diff
@@ -0,0 +1,271 @@
+#include "common/buffer/buffer_impl.h"
+
+#include "extensions/filters/network/zookeeper_proxy/zookeeper_decoder.h"
+#include "extensions/filters/network/zookeeper_proxy/zookeeper_filter.h"
+
+#include "test/mocks/network/mocks.h"
+#include "test/test_common/test_base.h"
+
+#include "gmock/gmock.h"
+
+using testing::NiceMock;
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace ZooKeeperProxy {
+
+class ZooKeeperFilterTest : public TestBase {
+public:
+  ZooKeeperFilterTest() { ENVOY_LOG_MISC(info, "test"); }
+
+  void initialize() {
+    config_ = std::make_shared<ZooKeeperFilterConfig>(stat_prefix_, scope_);
+    filter_ = std::make_unique<ZooKeeperFilter>(config_);
+    filter_->initializeReadFilterCallbacks(filter_callbacks_);
+  }
+
+  std::string encodeConnect(const bool readonly = false, const uint64_t zxid = 100,
+                            const uint32_t session_timeout = 10, const uint32_t session_id = 200,
+                            const std::string& passwd = "") const {
+    Buffer::OwnedImpl buffer;
+    const uint32_t message_size = readonly ? 28 + passwd.length() + 1 : 28 + passwd.length();
+
+    buffer.writeBEInt<uint32_t>(message_size);
+    buffer.writeBEInt<uint32_t>(0); // Protocol version.
+    buffer.writeBEInt<uint64_t>(zxid);
+    buffer.writeBEInt<uint32_t>(session_timeout);
+    buffer.writeBEInt<uint64_t>(session_id);
+    buffer.writeBEInt<uint32_t>(passwd.length());
+    buffer.add(passwd);
+
+    if (readonly) {
+      char readonly_flag = 0b1;
+      buffer.add(std::string(1, readonly_flag));
+    }
+
+    return buffer.toString();
+  }
+
+  std::string encodeBadMessage() const {
+    Buffer::OwnedImpl buffer;
+    // Bad length.
+    buffer.writeBEInt<uint32_t>(1);
+    // Trailing int.
+    buffer.writeBEInt<uint32_t>(3);
+
+    return buffer.toString();
+  }
+
+  std::string encodePing() const {
+    Buffer::OwnedImpl buffer;
+
+    buffer.writeBEInt<uint32_t>(8);
+    buffer.writeBEInt<int32_t>(PING_XID);
+
+    return buffer.toString();
+  }
+
+  std::string encodeAuth(const std::string& scheme) const {
+    Buffer::OwnedImpl buffer;
+
+    buffer.writeBEInt<uint32_t>(28 + scheme.length());
+    buffer.writeBEInt<int32_t>(AUTH_XID);
+    // Opcode.
+    buffer.writeBEInt<int32_t>(100);
+    // Type.
+    buffer.writeBEInt<int32_t>(0);
+    // Scheme.
+    buffer.writeBEInt<uint32_t>(scheme.length());
+    buffer.add(scheme);
+    // Credential.
+    buffer.writeBEInt<uint32_t>(6);
+    buffer.add("p@sswd");
+
+    return buffer.toString();
+  }
+
+  std::string encodeGetData(const std::string& path, const bool watch,
```

Comment:
I would just make this return the Buffer directly, converting to string doesn't seem to give you anything. Same for the other test helpers

Reply:
Good idea, done.